packages/ember-data/lib/system/relationships/ext.js

```diff
@@ -179,7 +178,7 @@ Model.reopenClass({
   */
   typeForRelationship: function(name) {
     var relationship = get(this, 'relationshipsByName').get(name);
-    return relationship && relationship.type;
+    return relationship && this.store.modelFor(relationship.type);
```

Comment:
this `store` shouldn't exist, as the goal of this work is to remove instance state from the factories. `this` here I believe is the factory.